File: apps/runtime/src/db/documents.rs
//! Board documents: loading a board's nodes, edges and whiteboard snapshot,
//! and saving them under an optimistic revision check.

use chrono::Utc;
use sqlx::{Row, SqlitePool};

use super::boards::get_board;
use super::validation::{validate_document, validate_viewport, validate_whiteboard};
use crate::{
    error::{AppError, AppResult},
    model::{BoardDocument, CanvasEdge, CanvasNode, Position, Size, Viewport},
};

// ---------------------------------------------------------------------------
// Board documents
// ---------------------------------------------------------------------------

pub async fn load_board(
    pool: &SqlitePool,
    workspace_id: &str,
    board_id: &str,
) -> AppResult<BoardDocument> {
    let board = get_board(pool, workspace_id, board_id).await?;

    let node_rows = sqlx::query(
        "SELECT id, board_id, type, title, color, x, y, width, height, collapsed, expanded_height, \
                parent_id, labels_json, note, data_json, created_at, updated_at \
         FROM nodes WHERE board_id = ? ORDER BY created_at",
    )
    .bind(&board.id)
    .fetch_all(pool)
    .await?;
    let nodes = node_rows
        .into_iter()
        .map(|row| {
            let data_json: String = row.try_get("data_json")?;
            let collapsed: i64 = row.try_get("collapsed")?;
            // Same reasoning as the board's kanban blob: a label list we cannot
            // decode is a missing chip, not a board that refuses to open.
            let labels_json: String = row.try_get("labels_json")?;
            Ok(CanvasNode {
                id: row.try_get("id")?,
                board_id: row.try_get("board_id")?,
                node_type: row.try_get("type")?,
                title: row.try_get("title")?,
                color: row.try_get("color")?,
                position: Position {
                    x: row.try_get("x")?,
                    y: row.try_get("y")?,
                },
                size: match (row.try_get("width")?, row.try_get("height")?) {
                    (Some(width), Some(height)) => Some(Size { width, height }),
                    _ => None,
                },
                collapsed: (collapsed != 0).then_some(true),
                expanded_height: row.try_get("expanded_height")?,
                parent_id: row.try_get("parent_id")?,
                labels: serde_json::from_str(&labels_json).unwrap_or_default(),
                note: row.try_get("note")?,
                data: serde_json::from_str(&data_json)
                    .map_err(|error| sqlx::Error::Decode(Box::new(error)))?,
                created_at: row.try_get("created_at")?,
                updated_at: row.try_get("updated_at")?,
            })
        })
        .collect::<Result<Vec<_>, sqlx::Error>>()?;

    let edge_rows = sqlx::query(
        "SELECT id, board_id, source_node_id, target_node_id, kind, created_at, updated_at \
         FROM edges WHERE board_id = ? ORDER BY created_at",
    )
    .bind(&board.id)
    .fetch_all(pool)
    .await?;
    let edges = edge_rows
        .into_iter()
        .map(|row| {
            Ok(CanvasEdge {
                id: row.try_get("id")?,
                board_id: row.try_get("board_id")?,
                source: row.try_get("source_node_id")?,
                target: row.try_get("target_node_id")?,
                kind: row.try_get("kind")?,
                created_at: row.try_get("created_at")?,
                updated_at: row.try_get("updated_at")?,
            })
        })
        .collect::<Result<Vec<_>, sqlx::Error>>()?;

    Ok(BoardDocument {
        board,
        nodes,
        edges,
    })
}

pub struct SaveBoardRequest<'a> {
    pub expected_updated_at: &'a str,
    pub nodes: &'a [CanvasNode],
    pub edges: &'a [CanvasEdge],
    pub viewport: Viewport,
    /// Omitting the drawing snapshot preserves the stored whiteboard.
    pub whiteboard: Option<&'a str>,
}

pub async fn save_board(
    pool: &SqlitePool,
    workspace_id: &str,
    board_id: &str,
    request: SaveBoardRequest<'_>,
) -> AppResult<BoardDocument> {
    let board = get_board(pool, workspace_id, board_id).await?;
    validate_document(&board.id, request.nodes, request.edges)?;
    validate_viewport(&request.viewport)?;

    let viewport_json = serde_json::to_string(&request.viewport)
        .map_err(|error| AppError::BadRequest(error.to_string()))?;
    let whiteboard = match request.whiteboard {
        Some(snapshot) => {
            validate_whiteboard(snapshot)?;
            snapshot
        }
        None => board.whiteboard.as_str(),
    };

    let mut transaction = pool.begin().await?;
    let next_updated_at = Utc::now().to_rfc3339();
    let updated = sqlx::query(
        "UPDATE boards SET updated_at = ?, viewport_json = ?, whiteboard_json = ? \
         WHERE id = ? AND updated_at = ?",
    )
    .bind(&next_updated_at)
    .bind(&viewport_json)
    .bind(whiteboard)
    .bind(&board.id)
    .bind(request.expected_updated_at)
    .execute(&mut *transaction)
    .await?;
    if updated.rows_affected() != 1 {
        return Err(AppError::Conflict(
            "Board changed since it was loaded; reload before saving".into(),
        ));
    }
    // A save is a diff, never a rewrite. `agent_mailbox.source_node_id` and
    // `.target_node_id` are `ON DELETE CASCADE` on `nodes`, so deleting and
    // re-inserting every row would drop every message on the board on every
    // autosave — a drag, a rename or a colour change included. Nodes the
    // document still has are updated in place; only the ones it dropped are
    // deleted, and those *should* take their mailbox with them.
    let stored_edge_ids: Vec<String> =
        sqlx::query_scalar("SELECT id FROM edges WHERE board_id = ?")
            .bind(&board.id)
            .fetch_all(&mut *transaction)
            .await?;
    let stored_node_ids: Vec<String> =
        sqlx::query_scalar("SELECT id FROM nodes WHERE board_id = ?")
            .bind(&board.id)
            .fetch_all(&mut *transaction)
            .await?;
    let kept_edge_ids = request
        .edges
        .iter()
        .map(|edge| edge.id.as_str())
        .collect::<std::collections::HashSet<_>>();
    let kept_node_ids = request
        .nodes
        .iter()
        .map(|node| node.id.as_str())
        .collect::<std::collections::HashSet<_>>();

    // Edges first: an edge the document dropped may point at a node it also
    // dropped, and `edges` is `ON DELETE CASCADE` on `nodes` too. Every edge
    // that stays has both endpoints in `request.nodes` (`validate_document`),
    // so nothing surviving the node deletes is left dangling.
    for id in stored_edge_ids
        .iter()
        .filter(|id| !kept_edge_ids.contains(id.as_str()))
    {
        sqlx::query("DELETE FROM edges WHERE id = ?")
            .bind(id)
            .execute(&mut *transaction)
            .await?;
    }
    for id in stored_node_ids
        .iter()
        .filter(|id| !kept_node_ids.contains(id.as_str()))
    {
        sqlx::query("DELETE FROM nodes WHERE id = ?")
            .bind(id)
            .execute(&mut *transaction)
            .await?;
    }

    for node in request.nodes {
        // The `WHERE` guard keeps an id owned by another board from being
        // moved onto this one: the plain INSERT used to fail on the primary
        // key, and silently stealing the row would be worse than either.
        let written = sqlx::query(
            "INSERT INTO nodes (id, board_id, type, title, color, x, y, width, height, collapsed, \
                                expanded_height, parent_id, labels_json, note, data_json, \
                                created_at, updated_at) \
             VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) \
             ON CONFLICT(id) DO UPDATE SET \
               type = excluded.type, title = excluded.title, color = excluded.color, \
               x = excluded.x, y = excluded.y, width = excluded.width, height = excluded.height, \
               collapsed = excluded.collapsed, expanded_height = excluded.expanded_height, \
               parent_id = excluded.parent_id, labels_json = excluded.labels_json, \
               note = excluded.note, data_json = excluded.data_json, \
               created_at = excluded.created_at, updated_at = excluded.updated_at \
             WHERE nodes.board_id = excluded.board_id",
        )
        .bind(&node.id)
        .bind(&node.board_id)
        .bind(&node.node_type)
        .bind(&node.title)
        .bind(&node.color)
        .bind(node.position.x)
        .bind(node.position.y)
        .bind(node.size.as_ref().map(|size| size.width))
        .bind(node.size.as_ref().map(|size| size.height))
        .bind(i64::from(node.collapsed.unwrap_or(false)))
        .bind(node.expanded_height)
        .bind(&node.parent_id)
        .bind(
            serde_json::to_string(&node.labels)
                .map_err(|error| AppError::BadRequest(error.to_string()))?,
        )
        .bind(&node.note)
        .bind(
            serde_json::to_string(&node.data)
                .map_err(|error| AppError::BadRequest(error.to_string()))?,
        )
        .bind(&node.created_at)
        .bind(&node.updated_at)
        .execute(&mut *transaction)
        .await?;
        if written.rows_affected() != 1 {
            return Err(AppError::BadRequest(
                "Board contains a node that belongs to another board".into(),
            ));
        }
    }
    for edge in request.edges {
        let written = sqlx::query(
            "INSERT INTO edges (id, board_id, source_node_id, target_node_id, kind, created_at, updated_at) \
             VALUES (?, ?, ?, ?, ?, ?, ?) \
             ON CONFLICT(id) DO UPDATE SET \
               source_node_id = excluded.source_node_id, \
               target_node_id = excluded.target_node_id, kind = excluded.kind, \
               created_at = excluded.created_at, updated_at = excluded.updated_at \
             WHERE edges.board_id = excluded.board_id",
        )
        .bind(&edge.id)
        .bind(&edge.board_id)
        .bind(&edge.source)
        .bind(&edge.target)
        .bind(&edge.kind)
        .bind(&edge.created_at)
        .bind(&edge.updated_at)
        .execute(&mut *transaction)
        .await?;
        if written.rows_affected() != 1 {
            return Err(AppError::BadRequest(
                "Board contains an edge that belongs to another board".into(),
            ));
        }
    }
    transaction.commit().await?;
    load_board(pool, workspace_id, board_id).await
}

File: apps/desktop/src/core/canvas/documents.ts
import type { DatabaseSync } from "node:sqlite";
import { badRequest, conflict, rfc3339 } from "../workspaces/support";
import { getBoard } from "./boards";
import type {
  BoardDocument,
  CanvasEdge,
  CanvasNode,
  SaveBoardRequest,
} from "./document-types";
import { forgetNodes } from "./orphans";
import {
  validateDocument,
  validateViewport,
  validateWhiteboard,
} from "./validation";

/**
 * Loading a board's nodes, edges and whiteboard snapshot, and saving them
 * under an optimistic revision check.
 *
 * A port of `apps/runtime/src/db/documents.rs`. Two things in it are contract
 * rather than implementation, and neither may drift:
 *
 *   * **The CAS.** `UPDATE … WHERE id = ? AND updated_at = ?` is the whole of
 *     it. A row count other than 1 is a 409, and the front end answers a 409
 *     by rebasing its unsaved edits onto a fresh load and trying again. The
 *     board's `updated_at` *is* its revision number.
 *   * **A save is a diff, never a rewrite.** `agent_mailbox` is
 *     `ON DELETE CASCADE` on `nodes`, so deleting and re-inserting every row
 *     would drop every message on the board on every autosave — a drag, a
 *     rename or a colour change included.
 */

interface NodeRow {
  id: string;
  board_id: string;
  type: string;
  title: string;
  color: string;
  x: number;
  y: number;
  width: number | null;
  height: number | null;
  collapsed: number;
  expanded_height: number | null;
  parent_id: string | null;
  labels_json: string;
  note: string;
  data_json: string;
  created_at: string;
  updated_at: string;
}

interface EdgeRow {
  id: string;
  board_id: string;
  source_node_id: string;
  target_node_id: string;
  kind: string;
  created_at: string;
  updated_at: string;
}

export function loadBoard(
  database: DatabaseSync,
  workspaceId: string,
  boardId: string,
): BoardDocument {
  const board = getBoard(database, workspaceId, boardId);
  const nodeRows = database
    .prepare(
      "SELECT id, board_id, type, title, color, x, y, width, height, collapsed, expanded_height, " +
        "parent_id, labels_json, note, data_json, created_at, updated_at " +
        "FROM nodes WHERE board_id = ? ORDER BY created_at",
    )
    .all(board.id) as unknown as NodeRow[];
  const edgeRows = database
    .prepare(
      "SELECT id, board_id, source_node_id, target_node_id, kind, created_at, updated_at " +
        "FROM edges WHERE board_id = ? ORDER BY created_at",
    )
    .all(board.id) as unknown as EdgeRow[];
  return {
    board,
    nodes: nodeRows.map(nodeFromRow),
    edges: edgeRows.map((row) => ({
      id: row.id,
      boardId: row.board_id,
      source: row.source_node_id,
      target: row.target_node_id,
      kind: row.kind,
      createdAt: row.created_at,
      updatedAt: row.updated_at,
    })),
  };
}

function nodeFromRow(row: NodeRow): CanvasNode {
  // A label list we cannot decode is a missing chip, not a board that refuses
  // to open — the same reasoning the Rust reader applies.
  let labels: string[] = [];
  try {
    const parsed = JSON.parse(row.labels_json) as unknown;
    if (Array.isArray(parsed)) {
      labels = parsed.filter(
        (entry): entry is string => typeof entry === "string",
      );
    }
  } catch {
    labels = [];
  }
  const node: Record<string, unknown> = {
    id: row.id,
    boardId: row.board_id,
    type: row.type,
    title: row.title,
    color: row.color,
    position: { x: row.x, y: row.y },
    labels,
    note: row.note,
    data: JSON.parse(row.data_json) as unknown,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
  if (row.width !== null && row.height !== null) {
    node.size = { width: row.width, height: row.height };
  }
  // `skip_serializing_if = "Option::is_none"` plus `(collapsed != 0).then_some(true)`:
  // a collapsed node carries `true`, an expanded one carries no key at all.
  if (Number(row.collapsed) !== 0) node.collapsed = true;
  if (row.expanded_height !== null) node.expandedHeight = row.expanded_height;
  if (row.parent_id !== null) node.parentId = row.parent_id;
  return orderNode(node);
}

/** Serde emits the struct's field order; a byte diff against Rust wants it. */
function orderNode(node: Record<string, unknown>): CanvasNode {
  const ordered: Record<string, unknown> = {};
  for (const key of [
    "id",
    "boardId",
    "type",
    "title",
    "color",
    "position",
    "size",
    "collapsed",
    "expandedHeight",
    "parentId",
    "labels",
    "note",
    "data",
    "createdAt",
    "updatedAt",
  ]) {
    if (key in node) ordered[key] = node[key];
  }
  return ordered as unknown as CanvasNode;
}

/**
 * The whole document, written under the caller's revision.
 *
 * Everything happens in one `BEGIN IMMEDIATE`: the CAS, the deletes, the
 * orphan cleanup and the upserts. A save that fails part way takes the
 * cleanup back with it — a node that still exists must never be missing its
 * status.
 */
export function saveBoard(
  database: DatabaseSync,
  workspaceId: string,
  boardId: string,
  request: SaveBoardRequest,
): BoardDocument {
  const board = getBoard(database, workspaceId, boardId);
  validateDocument(board.id, request.nodes, request.edges);
  validateViewport(request.viewport);

  const viewportJson = JSON.stringify({
    x: request.viewport.x,
    y: request.viewport.y,
    zoom: request.viewport.zoom,
  });
  let whiteboard = board.whiteboard;
  if (request.whiteboard !== undefined) {
    validateWhiteboard(request.whiteboard);
    whiteboard = request.whiteboard;
  }

  database.exec("BEGIN IMMEDIATE");
  try {
    const nextUpdatedAt = rfc3339();
    const updated = database
      .prepare(
        "UPDATE boards SET updated_at = ?, viewport_json = ?, whiteboard_json = ? " +
          "WHERE id = ? AND updated_at = ?",
      )
      .run(
        nextUpdatedAt,
        viewportJson,
        whiteboard,
        board.id,
        request.expectedUpdatedAt,
      );
    if (updated.changes !== 1) {
      throw conflict("Board changed since it was loaded; reload before saving");
    }

    const storedEdgeIds = (
      database
        .prepare("SELECT id FROM edges WHERE board_id = ?")
        .all(board.id) as unknown as { id: string }[]
    ).map((row) => row.id);
    const storedNodeIds = (
      database
        .prepare("SELECT id FROM nodes WHERE board_id = ?")
        .all(board.id) as unknown as { id: string }[]
    ).map((row) => row.id);
    const keptEdgeIds = new Set(request.edges.map((edge) => edge.id));
    const keptNodeIds = new Set(request.nodes.map((node) => node.id));

    // Edges first: an edge the document dropped may point at a node it also
    // dropped, and `edges` is `ON DELETE CASCADE` on `nodes` too. Every edge
    // that stays has both endpoints in `request.nodes` (`validateDocument`),
    // so nothing surviving the node deletes is left dangling.
    const deleteEdge = database.prepare("DELETE FROM edges WHERE id = ?");
    for (const id of storedEdgeIds) {
      if (!keptEdgeIds.has(id)) deleteEdge.run(id);
    }
    const droppedNodeIds = storedNodeIds.filter((id) => !keptNodeIds.has(id));
    const deleteNode = database.prepare("DELETE FROM nodes WHERE id = ?");
    for (const id of droppedNodeIds) deleteNode.run(id);
    forgetNodes(database, droppedNodeIds);

    // The `WHERE` guard keeps an id owned by another board from being moved
    // onto this one: a plain INSERT would fail on the primary key, and
    // silently stealing the row would be worse than either.
    const upsertNode = database.prepare(
      "INSERT INTO nodes (id, board_id, type, title, color, x, y, width, height, collapsed, " +
        "expanded_height, parent_id, labels_json, note, data_json, created_at, updated_at) " +
        "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) " +
        "ON CONFLICT(id) DO UPDATE SET " +
        "type = excluded.type, title = excluded.title, color = excluded.color, " +
        "x = excluded.x, y = excluded.y, width = excluded.width, height = excluded.height, " +
        "collapsed = excluded.collapsed, expanded_height = excluded.expanded_height, " +
        "parent_id = excluded.parent_id, labels_json = excluded.labels_json, " +
        "note = excluded.note, data_json = excluded.data_json, " +
        "created_at = excluded.created_at, updated_at = excluded.updated_at " +
        "WHERE nodes.board_id = excluded.board_id",
    );
    for (const node of request.nodes) {
      const written = upsertNode.run(
        node.id,
        node.boardId,
        node.type,
        node.title,
        node.color,
        node.position.x,
        node.position.y,
        node.size?.width ?? null,
        node.size?.height ?? null,
        node.collapsed === true ? 1 : 0,
        node.expandedHeight ?? null,
        node.parentId ?? null,
        JSON.stringify(node.labels),
        node.note,
        JSON.stringify(node.data),
        node.createdAt,
        node.updatedAt,
      );
      if (written.changes !== 1) {
        throw badRequest("Board contains a node that belongs to another board");
      }
    }
    const upsertEdge = database.prepare(
      "INSERT INTO edges (id, board_id, source_node_id, target_node_id, kind, created_at, updated_at) " +
        "VALUES (?, ?, ?, ?, ?, ?, ?) " +
        "ON CONFLICT(id) DO UPDATE SET " +
        "source_node_id = excluded.source_node_id, " +
        "target_node_id = excluded.target_node_id, kind = excluded.kind, " +
        "created_at = excluded.created_at, updated_at = excluded.updated_at " +
        "WHERE edges.board_id = excluded.board_id",
    );
    for (const edge of request.edges) {
      const written = upsertEdge.run(
        edge.id,
        edge.boardId,
        edge.source,
        edge.target,
        edge.kind,
        edge.createdAt,
        edge.updatedAt,
      );
      if (written.changes !== 1) {
        throw badRequest(
          "Board contains an edge that belongs to another board",
        );
      }
    }
    database.exec("COMMIT");
  } catch (error) {
    database.exec("ROLLBACK");
    throw error;
  }
  return loadBoard(database, workspaceId, boardId);
}

export type { BoardDocument, CanvasEdge, CanvasNode, SaveBoardRequest };
